Please implement Method DatabaseMetaData.getTypeInfo in JDBC Driver

Hi guys,
Looks like this method has not been implemented as of yet.  Third party products such as DbVisualizer however do make use of it - and it does sound kinda important. :)

thank you!


Here's my stack trace if its any assistance:

2017-02-09 21:34:26.185 INFO    36 [ExecutorRunner-pool-3-thread-1 - L.ￄﾁ] Exception while caching data types for database: presto-160-prod
com.onseven.dbvis.K.B.Q: com.facebook.presto.jdbc.NotImplementedException: Method DatabaseMetaData.getTypeInfo is not yet implemented
    at com.onseven.dbvis.K.B.P.ￄﾡ(Z:1111)
    at com.onseven.dbvis.K.B._.ￄﾜ(Z:2717)
    at com.onseven.dbvis.K.B._.ￄﾎ(Z:2518)
    at com.onseven.dbvis.db.L.ￄﾕ(Z:3124)
    at com.onseven.dbvis.db.I.L.ￄﾡ(Z:1582)
    at com.onseven.dbvis.db.I.L.ￄﾁ(Z:1239)
    at com.onseven.dbvis.db.I.L.ￄﾃ(Z:1768)
    at com.onseven.dbvis.db.I.H.ￄﾳ(Z:2009)
    at com.onseven.dbvis.db.I.K.ￄﾳ(Z:235)
    at com.onseven.dbvis.db.I.D.ￄﾡ(Z:1756)
    at com.onseven.dbvis.K.B._.ￄﾜ(Z:2717)
    at com.onseven.dbvis.K.B.L.ￄﾂ(Z:1374)
    at com.onseven.dbvis.K.B.L.doInBackground(Z:1521)
    at javax.swing.SwingWorker$1.call(SwingWorker.java:295)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at javax.swing.SwingWorker.run(SwingWorker.java:334)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: com.facebook.presto.jdbc.NotImplementedException: Method DatabaseMetaData.getTypeInfo is not yet implemented
    at com.facebook.presto.jdbc.PrestoDatabaseMetaData.getTypeInfo(PrestoDatabaseMetaData.java:1055)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at com.onseven.dbvis.g.B.E.ￄﾁ(Z:482)
    at com.onseven.dbvis.g.B.F$A.call(Z:1369)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)



Here's some config info from my box:

2017-02-09 21:23:52.545 CONFIG  36 [AWT-EventQueue-0 - D.ￄﾁ] Product: DbVisualizer 9.5.6 [Build #2608]
2017-02-09 21:23:52.548 CONFIG  36 [AWT-EventQueue-0 - D.ￄﾁ] OS: Mac OS X
2017-02-09 21:23:52.548 CONFIG  36 [AWT-EventQueue-0 - D.ￄﾁ] OS Version: 10.12.1
2017-02-09 21:23:52.548 CONFIG  36 [AWT-EventQueue-0 - D.ￄﾁ] OS Arch: x86_64
2017-02-09 21:23:52.549 CONFIG  36 [AWT-EventQueue-0 - D.ￄﾁ] Java Version: 1.8.0_111
2017-02-09 21:23:52.549 CONFIG  36 [AWT-EventQueue-0 - D.ￄﾁ] Java VM: Java HotSpot(TM) 64-Bit Server VM
2017-02-09 21:23:52.549 CONFIG  36 [AWT-EventQueue-0 - D.ￄﾁ] Java Vendor: Oracle Corporation
2017-02-09 21:23:52.549 CONFIG  36 [AWT-EventQueue-0 - D.ￄﾁ] Java Home: /Library/Internet Plug-Ins/JavaAppletPlugin.plugin/Contents/Home
2017-02-09 21:23:52.549 CONFIG  36 [AWT-EventQueue-0 - D.ￄﾁ] DbVis Home: /Applications/DbVisualizer.app/Contents/java/app
2017-02-09 21:23:52.549 CONFIG  36 [AWT-EventQueue-0 - D.ￄﾁ] User Home: /Users/spragues
2017-02-09 21:23:52.550 CONFIG  36 [AWT-EventQueue-0 - D.ￄﾁ] PrefsDir: /Users/spragues/.dbvis

--- a/presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoDatabaseMetaData.java
+++ b/presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoDatabaseMetaData.java
@@ -1044,8 +1044,13 @@ public class PrestoDatabaseMetaData
     public ResultSet getTypeInfo()
             throws SQLException
     {
-        // TODO: implement this
-        throw new NotImplementedException("DatabaseMetaData", "getTypeInfo");
+        return select("" +
+                "SELECT TYPE_NAME, DATA_TYPE, PRECISION, LITERAL_PREFIX, LITERAL_SUFFIX,\n" +
+                "CREATE_PARAMS, NULLABLE, CASE_SENSITIVE, SEARCHABLE, UNSIGNED_ATTRIBUTE,\n" +
+                "FIXED_PREC_SCALE, AUTO_INCREMENT, LOCAL_TYPE_NAME, MINIMUM_SCALE, MAXIMUM_SCALE,\n" +
+                "SQL_DATA_TYPE, SQL_DATETIME_SUB, NUM_PREC_RADIX\n" +
+                "FROM system.jdbc.types\n" +
+                "ORDER BY DATA_TYPE");
     }
 
     @Override

--- a/presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestPrestoDatabaseMetaData.java
+++ b/presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestPrestoDatabaseMetaData.java
@@ -28,9 +28,11 @@ import java.sql.DatabaseMetaData;
 import java.sql.DriverManager;
 import java.sql.ResultSet;
 import java.sql.SQLException;
+import java.sql.Types;
 import java.util.Set;
 import java.util.concurrent.Callable;
 
+import static com.facebook.presto.jdbc.TestPrestoDriver.closeQuietly;
 import static com.facebook.presto.util.ImmutableCollectors.toImmutableSet;
 import static com.google.common.collect.Iterables.getOnlyElement;
 import static io.airlift.testing.Assertions.assertContains;
@@ -71,7 +73,7 @@ public class TestPrestoDatabaseMetaData
     public void tearDown()
             throws SQLException
     {
-        connection.close();
+        closeQuietly(connection);
     }
 
     @Test
@@ -92,6 +94,64 @@ public class TestPrestoDatabaseMetaData
         String query = getOnlyElement(queries);
 
         assertContains(query, "_t' ESCAPE '", "Metadata query does not contain ESCAPE");
+    }
+
+    @Test
+    public void testGetTypeInfo()
+            throws Exception
+    {
+        DatabaseMetaData metaData = connection.getMetaData();
+        ResultSet typeInfo = metaData.getTypeInfo();
+        while (typeInfo.next()) {
+            int jdbcType = typeInfo.getInt("DATA_TYPE");
+            switch (jdbcType) {
+                case Types.BIGINT:
+                    assertColumnSpec(typeInfo, Types.BIGINT, 19L, 10L, "bigint");
+                    break;
+                case Types.BOOLEAN:
+                    assertColumnSpec(typeInfo, Types.BOOLEAN, null, null, "boolean");
+                    break;
+                case Types.INTEGER:
+                    assertColumnSpec(typeInfo, Types.INTEGER, 10L, 10L, "integer");
+                    break;
+                case Types.DECIMAL:
+                    assertColumnSpec(typeInfo, Types.DECIMAL, 38L, 10L, "decimal");
+                    break;
+                case Types.VARCHAR:
+                    assertColumnSpec(typeInfo, Types.VARCHAR, null, null, "varchar");
+                    break;
+                case Types.TIMESTAMP:
+                    assertColumnSpec(typeInfo, Types.TIMESTAMP, 23L, null, "timestamp");
+                    break;
+                case Types.DOUBLE:
+                    assertColumnSpec(typeInfo, Types.DOUBLE, 53L, 2L, "double");
+                    break;
+            }
+        }
+    }
+
+    private static void assertColumnSpec(ResultSet rs, int dataType, Long precision, Long numPrecRadix, String typeName)
+            throws SQLException
+    {
+        String message = " of " + typeName + ": ";
+        assertEquals(rs.getObject("TYPE_NAME"), typeName, "TYPE_NAME" + message);
+        assertEquals(rs.getObject("DATA_TYPE"), (long) dataType, "DATA_TYPE" + message);
+        assertEquals(rs.getObject("PRECISION"), precision, "PRECISION" + message);
+        assertEquals(rs.getObject("LITERAL_PREFIX"), null, "LITERAL_PREFIX" + message);
+        assertEquals(rs.getObject("LITERAL_SUFFIX"), null, "LITERAL_SUFFIX" + message);
+        assertEquals(rs.getObject("CREATE_PARAMS"), null, "CREATE_PARAMS" + message);
+        assertEquals(rs.getObject("NULLABLE"), (long) DatabaseMetaData.typeNullable, "NULLABLE" + message);
+        assertEquals(rs.getObject("CASE_SENSITIVE"), false, "CASE_SENSITIVE" + message);
+        assertEquals(rs.getObject("SEARCHABLE"), (long) DatabaseMetaData.typeSearchable, "SEARCHABLE" + message);
+        assertEquals(rs.getObject("UNSIGNED_ATTRIBUTE"), null, "UNSIGNED_ATTRIBUTE" + message);
+        assertEquals(rs.getObject("FIXED_PREC_SCALE"), false, "FIXED_PREC_SCALE" + message);
+        assertEquals(rs.getObject("AUTO_INCREMENT"), null, "AUTO_INCREMENT" + message);
+        assertEquals(rs.getObject("LOCAL_TYPE_NAME"), null, "LOCAL_TYPE_NAME" + message);
+        assertEquals(rs.getObject("MINIMUM_SCALE"), 0L, "MINIMUM_SCALE" + message);
+        assertEquals(rs.getObject("MAXIMUM_SCALE"), 0L, "MAXIMUM_SCALE" + message);
+        assertEquals(rs.getObject("SQL_DATA_TYPE"), null, "SQL_DATA_TYPE" + message);
+        assertEquals(rs.getObject("SQL_DATETIME_SUB"), null, "SQL_DATETIME_SUB" + message);
+        assertEquals(rs.getObject("NUM_PREC_RADIX"), numPrecRadix, "NUM_PREC_RADIX" + message);
     }
 
     private Set<String> captureQueries(Callable<?> action)

--- a/presto-main/src/main/java/com/facebook/presto/connector/system/SystemConnectorModule.java
+++ b/presto-main/src/main/java/com/facebook/presto/connector/system/SystemConnectorModule.java
@@ -25,6 +25,7 @@ import com.facebook.presto.connector.system.jdbc.SuperTableJdbcTable;
 import com.facebook.presto.connector.system.jdbc.SuperTypeJdbcTable;
 import com.facebook.presto.connector.system.jdbc.TableJdbcTable;
 import com.facebook.presto.connector.system.jdbc.TableTypeJdbcTable;
+import com.facebook.presto.connector.system.jdbc.TypesJdbcTable;
 import com.facebook.presto.connector.system.jdbc.UdtJdbcTable;
 import com.facebook.presto.spi.SystemTable;
 import com.facebook.presto.spi.procedure.Procedure;
@@ -58,6 +59,7 @@ public class SystemConnectorModule
         globalTableBinder.addBinding().to(AttributeJdbcTable.class).in(Scopes.SINGLETON);
         globalTableBinder.addBinding().to(CatalogJdbcTable.class).in(Scopes.SINGLETON);
         globalTableBinder.addBinding().to(ColumnJdbcTable.class).in(Scopes.SINGLETON);
+        globalTableBinder.addBinding().to(TypesJdbcTable.class).in(Scopes.SINGLETON);
         globalTableBinder.addBinding().to(ProcedureColumnJdbcTable.class).in(Scopes.SINGLETON);
         globalTableBinder.addBinding().to(ProcedureJdbcTable.class).in(Scopes.SINGLETON);
         globalTableBinder.addBinding().to(PseudoColumnJdbcTable.class).in(Scopes.SINGLETON);

--- a/presto-main/src/main/java/com/facebook/presto/connector/system/jdbc/ColumnJdbcTable.java
+++ b/presto-main/src/main/java/com/facebook/presto/connector/system/jdbc/ColumnJdbcTable.java
@@ -166,7 +166,7 @@ public class ColumnJdbcTable
         }
     }
 
-    private static int jdbcDataType(Type type)
+    static int jdbcDataType(Type type)
     {
         if (type.equals(BOOLEAN)) {
             return Types.BOOLEAN;
@@ -222,7 +222,7 @@ public class ColumnJdbcTable
         return Types.JAVA_OBJECT;
     }
 
-    private static Integer columnSize(Type type)
+    static Integer columnSize(Type type)
     {
         if (type.equals(BIGINT)) {
             return 19;  // 2**63-1
@@ -295,7 +295,7 @@ public class ColumnJdbcTable
         return null;
     }
 
-    private static Integer numPrecRadix(Type type)
+    static Integer numPrecRadix(Type type)
     {
         if (type.equals(BIGINT) ||
                 type.equals(INTEGER) ||

--- a/presto-main/src/main/java/com/facebook/presto/connector/system/jdbc/TypesJdbcTable.java
+++ b/presto-main/src/main/java/com/facebook/presto/connector/system/jdbc/TypesJdbcTable.java
@@ -1,0 +1,142 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.connector.system.jdbc;
+
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.ConnectorTableMetadata;
+import com.facebook.presto.spi.InMemoryRecordSet;
+import com.facebook.presto.spi.InMemoryRecordSet.Builder;
+import com.facebook.presto.spi.RecordCursor;
+import com.facebook.presto.spi.SchemaTableName;
+import com.facebook.presto.spi.connector.ConnectorTransactionHandle;
+import com.facebook.presto.spi.predicate.TupleDomain;
+import com.facebook.presto.spi.type.ParametricType;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeManager;
+
+import javax.inject.Inject;
+
+import java.sql.DatabaseMetaData;
+import java.sql.Types;
+import java.util.Collection;
+
+import static com.facebook.presto.connector.system.jdbc.ColumnJdbcTable.columnSize;
+import static com.facebook.presto.connector.system.jdbc.ColumnJdbcTable.jdbcDataType;
+import static com.facebook.presto.connector.system.jdbc.ColumnJdbcTable.numPrecRadix;
+import static com.facebook.presto.metadata.MetadataUtil.TableMetadataBuilder.tableMetadataBuilder;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.spi.type.VarcharType.createUnboundedVarcharType;
+import static java.util.Objects.requireNonNull;
+
+public class TypesJdbcTable
+        extends JdbcTable
+{
+    public static final SchemaTableName NAME = new SchemaTableName("jdbc", "types");
+
+    public static final ConnectorTableMetadata METADATA = tableMetadataBuilder(NAME)
+            .column("type_name", createUnboundedVarcharType())
+            .column("data_type", BIGINT)
+            .column("precision", BIGINT)
+            .column("literal_prefix", createUnboundedVarcharType())
+            .column("literal_suffix", createUnboundedVarcharType())
+            .column("create_params", createUnboundedVarcharType())
+            .column("nullable", BIGINT)
+            .column("case_sensitive", BOOLEAN)
+            .column("searchable", BIGINT)
+            .column("unsigned_attribute", BOOLEAN)
+            .column("fixed_prec_scale", BOOLEAN)
+            .column("auto_increment", BOOLEAN)
+            .column("local_type_name", createUnboundedVarcharType())
+            .column("minimum_scale", BIGINT)
+            .column("maximum_scale", BIGINT)
+            .column("sql_data_type", BIGINT)
+            .column("sql_datetime_sub", BIGINT)
+            .column("num_prec_radix", BIGINT)
+            .build();
+
+    private final TypeManager typeManager;
+
+    @Inject
+    public TypesJdbcTable(TypeManager typeManager)
+    {
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+    }
+
+    @Override
+    public ConnectorTableMetadata getTableMetadata()
+    {
+        return METADATA;
+    }
+
+    @Override
+    public RecordCursor cursor(ConnectorTransactionHandle transactionHandle, ConnectorSession connectorSession, TupleDomain<Integer> constraint)
+    {
+        Builder table = InMemoryRecordSet.builder(METADATA);
+        for (Type type : typeManager.getTypes()) {
+            addTypeRow(table, type);
+        }
+        addParametricTypeRows(table, typeManager.getParametricTypes());
+        return table.build().cursor();
+    }
+
+    private static void addTypeRow(Builder builder, Type type)
+    {
+        builder.addRow(
+                type.getDisplayName(),
+                jdbcDataType(type),
+                columnSize(type),
+                null,
+                null,
+                null,
+                DatabaseMetaData.typeNullable,
+                false,
+                type.isComparable() ? DatabaseMetaData.typeSearchable : DatabaseMetaData.typePredNone,
+                null,
+                false,
+                null,
+                null,
+                0,
+                0,
+                null,
+                null,
+                numPrecRadix(type));
+    }
+
+    private static void addParametricTypeRows(Builder builder, Collection<ParametricType> types)
+    {
+        for (ParametricType type : types) {
+            String typeName = type.getName();
+            builder.addRow(
+                    typeName,
+                    typeName.equalsIgnoreCase("array") ? Types.ARRAY : Types.JAVA_OBJECT,
+                    null,
+                    null,
+                    null,
+                    null,
+                    DatabaseMetaData.typeNullable,
+                    false,
+                    DatabaseMetaData.typePredNone,
+                    null,
+                    false,
+                    null,
+                    null,
+                    0,
+                    0,
+                    null,
+                    null,
+                    null);
+        }
+    }
+}

--- a/presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java
+++ b/presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java
@@ -30,6 +30,7 @@ import javax.annotation.concurrent.ThreadSafe;
 import javax.inject.Inject;
 
 import java.util.ArrayList;
+import java.util.Collection;
 import java.util.List;
 import java.util.Locale;
 import java.util.Optional;
@@ -320,6 +321,12 @@ public final class TypeRegistry
         String name = parametricType.getName().toLowerCase(Locale.ENGLISH);
         checkArgument(!parametricTypes.containsKey(name), "Parametric type already registered: %s", name);
         parametricTypes.putIfAbsent(name, parametricType);
+    }
+
+    @Override
+    public Collection<ParametricType> getParametricTypes()
+    {
+        return ImmutableList.copyOf(parametricTypes.values());
     }
 
     /**

--- a/presto-spi/src/main/java/com/facebook/presto/spi/type/TypeManager.java
+++ b/presto-spi/src/main/java/com/facebook/presto/spi/type/TypeManager.java
@@ -13,6 +13,7 @@
  */
 package com.facebook.presto.spi.type;
 
+import java.util.Collection;
 import java.util.List;
 import java.util.Optional;
 
@@ -32,6 +33,11 @@ public interface TypeManager
      * Gets a list of all registered types.
      */
     List<Type> getTypes();
+
+    /**
+     * Gets all registered parametric types.
+     */
+    Collection<ParametricType> getParametricTypes();
 
     Optional<Type> getCommonSuperType(Type firstType, Type secondType);
 

--- a/presto-spi/src/test/java/com/facebook/presto/spi/type/TestingTypeManager.java
+++ b/presto-spi/src/test/java/com/facebook/presto/spi/type/TestingTypeManager.java
@@ -15,6 +15,7 @@ package com.facebook.presto.spi.type;
 
 import com.google.common.collect.ImmutableList;
 
+import java.util.Collection;
 import java.util.List;
 import java.util.Optional;
 
@@ -52,6 +53,12 @@ public class TestingTypeManager
     public List<Type> getTypes()
     {
         return ImmutableList.of(BOOLEAN, BIGINT, DOUBLE, VARCHAR, VARBINARY, TIMESTAMP, DATE, ID, HYPER_LOG_LOG);
+    }
+
+    @Override
+    public Collection<ParametricType> getParametricTypes()
+    {
+        return ImmutableList.of();
     }
 
     @Override

